<a href="https://colab.research.google.com/github/SunilY137/EDA-1/blob/main/Play_Store_App_Review_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <b> The Play Store apps data has enormous potential to drive app-making businesses to success. Actionable insights can be drawn for developers to work on and capture the Android market. </b>

## <b> Each app (row) has values for catergory, rating, size, and more. Another dataset contains customer reviews of the android apps.</b>

## <b> Explore and analyze the data to discover key factors responsible for app engagement and success. </b>

In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
from scipy import stats
import seaborn as sns
from google.colab import drive
drive.mount("/content/gdrive")

Mounted at /content/gdrive


In [2]:
df = pd.read_csv('/content/gdrive/MyDrive/EDA Project 1/Play Store Data.csv', dtype='str')


# 1 We first use head() method to get a quick look at the first 5 rows of data.
# Here we can make initial thoughts on what might need to be cleaned or manipulated.
# A info() and shape() will give us a little more information of the dataset. 
# The shape method reveals that we have 10841 rows and 13 columns

In [3]:
df.head()




App        Category Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN    4.1   
1                                Coloring book moana  ART_AND_DESIGN    3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN    4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN    4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN    4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [4]:
df.shape

(10841, 13)

In [20]:
df.describe

<bound method NDFrame.describe of                                                      app             category  \
0         PHOTO EDITOR & CANDY CAMERA & GRID & SCRAPBOOK       ART_AND_DESIGN   
1                                    COLORING BOOK MOANA       ART_AND_DESIGN   
2      U LAUNCHER LITE – FREE LIVE COOL THEMES, HIDE ...       ART_AND_DESIGN   
3                                  SKETCH - DRAW & PAINT       ART_AND_DESIGN   
4                  PIXEL DRAW - NUMBER ART COLORING BOOK       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   SYA9A MAROC - FR               FAMILY   
10837                   FR. MIKE SCHMITZ AUDIO TEACHINGS               FAMILY   
10838                             PARKINSON EXERCICES FR              MEDICAL   
10839                      THE SCP FOUNDATION DB FR NN5N  BOOKS_AND_REFERENCE   
10840      IHOROSCOPE - 2018 DAILY HOROSCOPE & ASTROLOGY            LIFESTY

In [5]:
df.size

140933

# 2. Data Cleaning Phase

2. Data Cleaning Phase
First of all, we can do some general tidying up of the column names and values.clean columns names and values

In [6]:
def clean_columns(df):
    df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')
    
clean_columns(df)

def clean_vals(df):
    for col in df.columns:
        df[col] = df[col].str.strip().str.upper().str.replace('\t|\n', '', regex=True)
     
clean_vals(df)

In [7]:
#check

df.head()

app        category rating  \
0     PHOTO EDITOR & CANDY CAMERA & GRID & SCRAPBOOK  ART_AND_DESIGN    4.1   
1                                COLORING BOOK MOANA  ART_AND_DESIGN    3.9   
2  U LAUNCHER LITE – FREE LIVE COOL THEMES, HIDE ...  ART_AND_DESIGN    4.7   
3                              SKETCH - DRAW & PAINT  ART_AND_DESIGN    4.5   
4              PIXEL DRAW - NUMBER ART COLORING BOOK  ART_AND_DESIGN    4.3   

  reviews  size     installs  type price content_rating  \
0     159   19M      10,000+  FREE     0       EVERYONE   
1     967   14M     500,000+  FREE     0       EVERYONE   
2   87510  8.7M   5,000,000+  FREE     0       EVERYONE   
3  215644   25M  50,000,000+  FREE     0           TEEN   
4     967  2.8M     100,000+  FREE     0       EVERYONE   

                      genres      last_updated         current_ver  \
0               ART & DESIGN   JANUARY 7, 2018               1.0.0   
1  ART & DESIGN;PRETEND PLAY  JANUARY 15, 2018               2.0.0   
2               ART & DESIGN    AUGUST 1, 2018               1.2.4   
3               ART & DESIGN      JUNE 8, 2018  VARIES WITH DEVICE   
4    ART & DESIGN;CREATIVITY     JUNE 20, 2018                 1.1   

    android_ver  
0  4.0.3 AND UP  
1  4.0.3 AND UP  
2  4.0.3 AND UP  
3    4.2 AND UP  
4    4.4 AND UP

# Size
We need to see what letters there are in the size column so that we can normalize the data.

We know that 1 MB = 1024 KB.

print(df['size'].str.replace('\d+', '').unique())
['M' '.M' 'VARIES WITH DEVICE' 'K' '.K' ',+']

  """Entry point for launching an IPython kernel.

   From the code above, we can see that we have our size data in the form of megabytes, kilobytes and unknown.

In [8]:
df.loc[df['size'].str.contains('M'),].head()

app        category rating  \
0     PHOTO EDITOR & CANDY CAMERA & GRID & SCRAPBOOK  ART_AND_DESIGN    4.1   
1                                COLORING BOOK MOANA  ART_AND_DESIGN    3.9   
2  U LAUNCHER LITE – FREE LIVE COOL THEMES, HIDE ...  ART_AND_DESIGN    4.7   
3                              SKETCH - DRAW & PAINT  ART_AND_DESIGN    4.5   
4              PIXEL DRAW - NUMBER ART COLORING BOOK  ART_AND_DESIGN    4.3   

  reviews  size     installs  type price content_rating  \
0     159   19M      10,000+  FREE     0       EVERYONE   
1     967   14M     500,000+  FREE     0       EVERYONE   
2   87510  8.7M   5,000,000+  FREE     0       EVERYONE   
3  215644   25M  50,000,000+  FREE     0           TEEN   
4     967  2.8M     100,000+  FREE     0       EVERYONE   

                      genres      last_updated         current_ver  \
0               ART & DESIGN   JANUARY 7, 2018               1.0.0   
1  ART & DESIGN;PRETEND PLAY  JANUARY 15, 2018               2.0.0   
2               ART & DESIGN    AUGUST 1, 2018               1.2.4   
3               ART & DESIGN      JUNE 8, 2018  VARIES WITH DEVICE   
4    ART & DESIGN;CREATIVITY     JUNE 20, 2018                 1.1   

    android_ver  
0  4.0.3 AND UP  
1  4.0.3 AND UP  
2  4.0.3 AND UP  
3    4.2 AND UP  
4    4.4 AND UP

In [9]:
# converting mb to kb

def convert_mb_kb(df):
    df.loc[df['size'].str.contains('M'),'size'] = df.loc[df['size'].str.contains('M'),'size'].str.rstrip('M').astype(float) * 1024

convert_mb_kb(df)

In [10]:
df.head()

app        category rating  \
0     PHOTO EDITOR & CANDY CAMERA & GRID & SCRAPBOOK  ART_AND_DESIGN    4.1   
1                                COLORING BOOK MOANA  ART_AND_DESIGN    3.9   
2  U LAUNCHER LITE – FREE LIVE COOL THEMES, HIDE ...  ART_AND_DESIGN    4.7   
3                              SKETCH - DRAW & PAINT  ART_AND_DESIGN    4.5   
4              PIXEL DRAW - NUMBER ART COLORING BOOK  ART_AND_DESIGN    4.3   

  reviews     size     installs  type price content_rating  \
0     159  19456.0      10,000+  FREE     0       EVERYONE   
1     967  14336.0     500,000+  FREE     0       EVERYONE   
2   87510   8908.8   5,000,000+  FREE     0       EVERYONE   
3  215644  25600.0  50,000,000+  FREE     0           TEEN   
4     967   2867.2     100,000+  FREE     0       EVERYONE   

                      genres      last_updated         current_ver  \
0               ART & DESIGN   JANUARY 7, 2018               1.0.0   
1  ART & DESIGN;PRETEND PLAY  JANUARY 15, 2018               2.0.0   
2               ART & DESIGN    AUGUST 1, 2018               1.2.4   
3               ART & DESIGN      JUNE 8, 2018  VARIES WITH DEVICE   
4    ART & DESIGN;CREATIVITY     JUNE 20, 2018                 1.1   

    android_ver  
0  4.0.3 AND UP  
1  4.0.3 AND UP  
2  4.0.3 AND UP  
3    4.2 AND UP  
4    4.4 AND UP

In [11]:
# Lets look at all our unique letters now

print(df['size'].astype(str).str.replace('\d+', '').unique())

['.' 'VARIES WITH DEVICE' 'K' '.K' ',+']


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  This is separate from the ipykernel package so we can avoid doing imports until


Now we can simply remove the K and change the size column to say "size (in kb)" for clarity,

In [12]:
df.loc[df['size'].astype(str).str.contains('K'), 'size'] = df.loc[df['size'].astype(str).str.contains('K'), 'size'].str.replace('K', '')

In [13]:
# we can see that there are now no more values containing the letter K
df.loc[df['size'].astype(str).str.contains('K'), 'size']



Series([], Name: size, dtype: object)

In [14]:
# adding clarify to the size column

df.rename(columns = {'size': 'size(kb)'}, inplace=True)

In [15]:
# check

df.columns

Index(['app', 'category', 'rating', 'reviews', 'size(kb)', 'installs', 'type',
       'price', 'content_rating', 'genres', 'last_updated', 'current_ver',
       'android_ver'],
      dtype='object')

We know that "VARIES WITH DEVICE" means unknown, or np.nan. So lets change replace those values.

In [16]:
df['size(kb)'].replace('VARIES WITH DEVICE', np.nan, inplace=True)

In [17]:
df.loc[df['size(kb)'].isnull(),]

app             category  \
37                                    FLOOR PLAN CREATOR       ART_AND_DESIGN   
42                            TEXTGRAM - WRITE ON PHOTOS       ART_AND_DESIGN   
52                         USED CARS AND TRUCKS FOR SALE    AUTO_AND_VEHICLES   
67                                    ULYSSE SPEEDOMETER    AUTO_AND_VEHICLES   
68                                                REPUVE    AUTO_AND_VEHICLES   
...                                                  ...                  ...   
10713  MY EARTHQUAKE ALERTS - US & WORLDWIDE EARTHQUAKES              WEATHER   
10725                                          POSTA APP  MAPS_AND_NAVIGATION   
10765                    CHAT FOR STRANGERS - VIDEO CHAT               SOCIAL   
10826          FRIM: GET NEW FRIENDS ON LOCAL CHAT ROOMS               SOCIAL   
10839                      THE SCP FOUNDATION DB FR NN5N  BOOKS_AND_REFERENCE   

      rating reviews size(kb)     installs  type price content_rating  \
37       4.1   36639      NaN   5,000,000+  FREE     0       EVERYONE   
42       4.4  295221      NaN  10,000,000+  FREE     0       EVERYONE   
52       4.6   17057      NaN   1,000,000+  FREE     0       EVERYONE   
67       4.3   40211      NaN   5,000,000+  FREE     0       EVERYONE   
68       3.9     356      NaN     100,000+  FREE     0       EVERYONE   
...      ...     ...      ...          ...   ...   ...            ...   
10713    4.4    3471      NaN     100,000+  FREE     0       EVERYONE   
10725    3.6       8      NaN       1,000+  FREE     0       EVERYONE   
10765    3.4     622      NaN     100,000+  FREE     0     MATURE 17+   
10826    4.0   88486      NaN   5,000,000+  FREE     0     MATURE 17+   
10839    4.5     114      NaN       1,000+  FREE     0     MATURE 17+   

                  genres        last_updated         current_ver  \
37          ART & DESIGN       JULY 14, 2018  VARIES WITH DEVICE   
42          ART & DESIGN       JULY 30, 2018  VARIES WITH DEVICE   
52       AUTO & VEHICLES       JULY 30, 2018  VARIES WITH DEVICE   
67       AUTO & VEHICLES       JULY 30, 2018  VARIES WITH DEVICE   
68       AUTO & VEHICLES        MAY 25, 2018  VARIES WITH DEVICE   
...                  ...                 ...                 ...   
10713            WEATHER       JULY 24, 2018  VARIES WITH DEVICE   
10725  MAPS & NAVIGATION  SEPTEMBER 27, 2017  VARIES WITH DEVICE   
10765             SOCIAL        MAY 23, 2018  VARIES WITH DEVICE   
10826             SOCIAL      MARCH 23, 2018  VARIES WITH DEVICE   
10839  BOOKS & REFERENCE    JANUARY 19, 2015  VARIES WITH DEVICE   

              android_ver  
37           2.3.3 AND UP  
42     VARIES WITH DEVICE  
52     VARIES WITH DEVICE  
67     VARIES WITH DEVICE  
68     VARIES WITH DEVICE  
...                   ...  
10713  VARIES WITH DEVICE  
10725          4.4 AND UP  
10765  VARIES WITH DEVICE  
10826  VARIES WITH DEVICE  
10839  VARIES WITH DEVICE  

[1695 rows x 13 columns]

From this we can see that there are 1695 rows that have NaN values and get a general idea of the context of those values.

In [21]:
def chk_null_sum(df, col):
    return df[col].isnull().sum()

chk_null_sum(df, 'size(kb)')

1695

In [22]:
# remove the "+"

def remove_plus(df, col):
    df[col] = df[col].astype(str).str.replace('+', '', regex=False)

remove_plus(df, 'size(kb)')

In [23]:
# check

df.loc[df['size(kb)'].astype(str).str.contains('+', regex=False),]

Empty DataFrame
Columns: [app, category, rating, reviews, size(kb), installs, type, price, content_rating, genres, last_updated, current_ver, android_ver]
Index: []

In [24]:
# removing commas

def remove_commas(df, col):
    df[col] = df[col].str.replace(',', '')

remove_commas(df, 'size(kb)')

In [25]:
# convert column size data type to numeric

def convert_to_numeric(df, col):
    df[col] = pd.to_numeric(df[col], errors='coerce')

convert_to_numeric(df, 'size(kb)')

I am using coerce here because I know that anything being coerced is already a nan value.


NaN values in size column
To deal with these missing values, we can fill them with the mean value, grouped by category, so that it does not affect our standard deviation.

In [26]:
df['size(kb)'] = df['size(kb)'].astype(float).fillna(df.groupby('category')['size(kb)'].transform('mean'))

In [27]:
# check

df.isnull().sum()

app                  0
category             0
rating            1474
reviews              0
size(kb)             0
installs             0
type                 1
price                0
content_rating       1
genres               0
last_updated         0
current_ver          8
android_ver          3
dtype: int64

## Last Updated
For this analysis, I only want the year so lets remove everything before the comma in the last_updated row.



In [28]:
df['last_updated'].unique()

array(['JANUARY 7, 2018', 'JANUARY 15, 2018', 'AUGUST 1, 2018', ...,
       'JANUARY 20, 2014', 'FEBRUARY 16, 2014', 'MARCH 23, 2014'],
      dtype=object)

In [29]:
cond = df.loc[df['last_updated'] == '1.0.19',].index

Its looks like this row has been currupted, so we will remove it from the dataframe.

In [30]:
df.drop(cond, inplace=True)

In [31]:
def convert_datetime(df, col):
    df[col] = pd.to_datetime(df[col])
    
def extract_year_dt(df, col):
    df[col] = df[col].dt.year
    
    
convert_datetime(df, 'last_updated')
extract_year_dt(df, 'last_updated')

In [32]:
# check

print(df['last_updated'])

0        2018
1        2018
2        2018
3        2018
4        2018
         ... 
10836    2017
10837    2018
10838    2017
10839    2015
10840    2018
Name: last_updated, Length: 10840, dtype: int64


## Installs
Lets remove any non numeric digits and convert the datatype to numeric.

In [33]:
df['installs'].unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
       '10+', '1+', '5+', '0+', '0'], dtype=object)

In [34]:
remove_plus(df, 'installs')
remove_commas(df, 'installs')

In [35]:
def to_numeric(df, col):
    return pd.to_numeric(df[col])

In [36]:
df['installs'] = to_numeric(df, 'installs')

In [38]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   app             10840 non-null  object 
 1   category        10840 non-null  object 
 2   rating          9366 non-null   object 
 3   reviews         10840 non-null  object 
 4   size(kb)        10840 non-null  float64
 5   installs        10840 non-null  int64  
 6   type            10839 non-null  object 
 7   price           10840 non-null  object 
 8   content_rating  10840 non-null  object 
 9   genres          10840 non-null  object 
 10  last_updated    10840 non-null  int64  
 11  current_ver     10832 non-null  object 
 12  android_ver     10838 non-null  object 
dtypes: float64(1), int64(2), object(10)
memory usage: 1.2+ MB


## Current_ver and Android_ver
We dont need current_ver or android_ver for this anlalysis so we will remove those.

In [39]:
df = df.drop(['current_ver', 'android_ver'], axis=1)

In [40]:
df.head()

app        category rating  \
0     PHOTO EDITOR & CANDY CAMERA & GRID & SCRAPBOOK  ART_AND_DESIGN    4.1   
1                                COLORING BOOK MOANA  ART_AND_DESIGN    3.9   
2  U LAUNCHER LITE – FREE LIVE COOL THEMES, HIDE ...  ART_AND_DESIGN    4.7   
3                              SKETCH - DRAW & PAINT  ART_AND_DESIGN    4.5   
4              PIXEL DRAW - NUMBER ART COLORING BOOK  ART_AND_DESIGN    4.3   

  reviews  size(kb)  installs  type price content_rating  \
0     159   19456.0     10000  FREE     0       EVERYONE   
1     967   14336.0    500000  FREE     0       EVERYONE   
2   87510    8908.8   5000000  FREE     0       EVERYONE   
3  215644   25600.0  50000000  FREE     0           TEEN   
4     967    2867.2    100000  FREE     0       EVERYONE   

                      genres  last_updated  
0               ART & DESIGN          2018  
1  ART & DESIGN;PRETEND PLAY          2018  
2               ART & DESIGN          2018  
3               ART & DESIGN          2018  
4    ART & DESIGN;CREATIVITY          2018